README.md
# `rtsVis`
[![CRAN version](https://www.r-pkg.org/badges/version/rtsVis)](https://CRAN.R-project.org/package=rtsVis)
[![CRAN downloads](https://cranlogs.r-pkg.org/badges/last-month/rtsVis?color=brightgreen)](https://CRAN.R-project.org/package=rtsVis)
[![CRAN checks](https://cranchecks.info/badges/summary/rtsVis)](https://cran.r-project.org/web/checks/check_results_rtsVis.html)
[![R-CMD-check](https://github.com/JohMast/rtsVis/workflows/R-CMD-check/badge.svg)](https://github.com/JohMast/rtsVis/actions)


A lightweight `R` package to visualize large raster time series, building on a fast temporal interpolation core.
rtsVis is linked to the <a href="https://github.com/16EAGLE/moveVis">`moveVis`</a> package and their joint use is recommended.

<img align="right" src="https://github.com/JohMast/rtsVis_demo/blob/main/Images/rtsVis_Logo.png" width="130" height="150" />

## Installation

The released version of rtsVis can be installed from CRAN:
```r
install.packages("rtsVis")
```


The latest version of the package can be installed from github. 

```r
devtools::install_github("JohMast/rtsVis")
```

## Concepts


rtsVis operates on lists of objects:

- Lists of <a href="https://CRAN.R-project.org/package=raster">`raster`</a> stacks 
- Lists of frames (<a href="https://ggplot2.tidyverse.org/">`ggplot`</a> objects of rasters or charts)

To process those lists in a pipeline, we recommend pipes such as provided by <a href="https://magrittr.tidyverse.org/">`magrittr`</a>.

- Positions of points or polygons, which can be added to the frames or be used to extract values from the time series and create animated charts.


## Functions
### Preparing rasters
* `ts_raster` Assemble/interpolate a raster time series.
* `ts_fill_na` Fill NA values in a raster time series
### Creating Frames
* `ts_flow_frames` 	Create a series of charts of a raster time series.
* `ts_makeframes` 	Create spatial ggplots of a raster time series.
* `ts_add_positions_to_frames` Add points, coordinates, or polygons to a list of spatial plots.

## Example
In this example, we use a MODIS time series <a href="https://td-host.rz.unibw-muenchen.de/btnncfkb/public/Wz3fRVuC?k=1CyESW64ZFybfaSRXaMoiCPCO6qNOx8ctc5FCGg1-Pw">` (download here) `</a>to create typical visualisations which highlight vegetation dynamics on the African continent.

### Part 1: Preparing the Data

We read our inputs:

* A list of **MODIS raster files** in tif format, which contain 4 bands (Red, NIR, Blue, Green). These are monthly composites of the MCD43A4.006 product.
* **points** in gpkg format, which can be visualized and for which statistics can be extracted. These correspond to various places in and around Africa.

Optional, but useful to enhance the visualizations:

* an **outline** of the continent Africa.
* a global **hillshade** raster which can be applied to the spatial plots.

Some functions in `rtsVis` require that timestamps for the rasters are set. Often, these come with the metadata or can be derived from the filename. They can also be manually set. In this example, we have monthly medians, and no true acquisition time. Therefore, we manually create a series of dates.
In addition to the input times, we also set output times - for these, output dates will be created. We simply create a second, denser series of dates. Having a denser series will smooth the animation.

If images from the input series are missing, it is not an issue, at least not from the technical side of things. Images for `out_times` will be interpolated regardless of the regularity of the input data. To illustrate this, we remove one of the input images. 

``` r
library(raster)
library(sf)
library(rtsVis)
library(RStoolbox)  
library(tidyverse)

# Load the inputs

modis_tifs <- list.files("Beispieldaten/MODIS/Africa_MODIS_2017-2020/",full.names = T,pattern=".tif$")
points <- st_read("Beispieldaten/Ancillary/Africa/Africa_places.gpkg")
outline <- st_read("Beispieldaten/Ancillary/Outline_continents_africa.gpkg") 
hillshade <- raster("Beispieldaten/Ancillary/SR_LR/SR_LR.tif") %>% readAll()


#Create in-times and out-times
in_dates <- as.POSIXct(seq.Date(as.Date("2017-01-15"),
                                as.Date("2020-12-15"),
                                length.out = length(modis_filled)))
                                
out_dates <-seq.POSIXt(from = in_dates[1],
                       to = in_dates[length(in_dates)],
                       length.out = length(in_dates)*4 )
                       
# simulate a missing image
in_dates <- in_dates[-13]
modis_tifs <- modis_tifs[-13]

```
### Part 2: Preparing the Rasters

For the preparation of the rasters, we use three functions:

* `stack` A <a href="https://CRAN.R-project.org/package=raster">`raster`</a> function which we use with `lapply` to load the tifs into a list.
* `ts_fill_na` We fill, wherever possible, missing values using simple temporal interpolation. This is not strictly necessary, but improves the visualization.
* `ts_raster` We assemble a full time series. This will interpolate additional missing frames for every desired output date and can take a long time for large time-series. 

Often, data requires additional preprocessing steps, such as reprojection, cropping, or masking.
These can be applied with `lapply` to retain the list format.

`ts_raster` returns a list of interpolated rasters, one for each desired output date, with timestamps attached.


``` r
#Read the images as stacks
modis_tifs_stacks <- modis_tifs %>%  lapply(stack) 

#fill NAs
modis_filled <- ts_fill_na(modis_tifs_stacks)

modis_ts <- ts_raster(
  r_list = modis_filled,
  r_times = in_dates,
  out_times = out_dates,
  fade_raster = T) 

```

### Part 3: Creating Basic RGB animations

In this step, `ts_makeframes` is used to create a list of frames (<a href="https://ggplot2.tidyverse.org/">`ggplot`</a> objects) from the time series (<a href="https://CRAN.R-project.org/package=raster">`raster`</a> objects).
We also add a outline to the plot area. This is one example of adding a *position* to a time series.
Using pipes is not necessary, but improves readability greatly.  


``` r

#create the frames from the raster
modis_frames_RGB <-  
  ts_makeframes(x_list = modis_ts,
                l_indices = c(1,4,3),  # MODIS bands are Red, NIR, Blue, Green
                minq = 0.0)            # adjust the stretch slightly

# Add the outline of the continent for visual clarity
modis_frames_RGB_ol <- 
  modis_frames_RGB %>% 
  ts_add_positions_to_frames(
    positions = outline,
    psize = 1) 

# Use moveVis utility to animate the frames into a gif
moveVis::animate_frames(modis_frames_RGB_ol,
                        overwrite = TRUE,
                        out_file = "modis_frames_RGB_ol.gif",
                        height=300,
                        width=300,
                        fps = 10,
                        res=75)
``` 

<p align="center"><img src="https://github.com/JohMast/rtsVis_demo/blob/main/Images/modis_frames_RGB_ol.gif"></p>

### Part 4: Customizing frames

The animation created suggests notable vegetation dynamics. An easy way to highlight this is the NDVI.

We calculate the NDVI using the `overlay` function provided by the <a href="https://CRAN.R-project.org/package=raster">`raster`</a> package, and reattach the timestamps using `ts_copy_frametimes`. Thereby we receive a second time series with a single layer per timestep. Thus, we do not create RGB frames, but gradient frames. 

Note that the individual frames are simply <a href="https://ggplot2.tidyverse.org/">`ggplot`</a> objects. Hence, <a href="https://github.com/16EAGLE/moveVis">`moveVis`</a>  functions and <a href="https://ggplot2.tidyverse.org/">`ggplot`</a> functions can be easily applied to the created frames using `lapply`. 



``` r

#calculate the NDVI per image
modis_ndvi <- modis_ts %>% lapply(function(x){
  ndvi <- overlay(x[[2]], x[[1]], fun=function(nir,red){(nir-red)/(nir+red)})
}) %>% rtsVis::ts_copy_frametimes(modis_ts)

# Create the frames from the NDVI raster
modis_ndvi_fr <-  
  ts_makeframes(x_list = modis_ndvi,
                hillshade = hillshade,
                r_type = "gradient")

# The frames can be adjusted like any ggplot 
modis_ndvi_fr_styled <- modis_ndvi_fr %>% lapply(FUN=function(x){
  x+scale_fill_gradient2("NDVI",
                         low = "red",
                         mid="yellow",
                         high = "green4",
                         midpoint=0.3,
                         limits=c(-0,1),
                         na.value = NA)+
    theme_bw() +
    xlab("Longitude")+
    ylab("Latitude")+
    ggtitle("Seasonality of NDVI in Africa",
            "Derived from MCD43A4 -- Monthly Composite 2015-2020 -- Projection: WGS 84 (EPSG 4326)")
}) %>%
  
  # packages like moveVis provide additional mapping functions
  moveVis::add_northarrow(colour = "black", position = "bottomleft") %>% 
  moveVis::add_progress()

```


### Part 5: Utilizing positions

`rtsVis` uses vector objects (positions) in two different ways:

* To **enhance** spatial frames, by adding noteworthy positions, such as test sites.
* To **create** flow frames, which are animated plots which visualize the data that is extracted under the positions.

Sometimes, adding a position is for visual appeal only, as we do here by adding an outline to the continents. 
But often, it makes sense to use the two together. In this example, we use (`ts_add_positions_to_frames`) to add several locations as points to our spatial frames. Subsequently, we extract values in a buffer around these locations and create a line chart that visualizes the development of these values over time.


``` r
#reduce the number of points slightly

modis_ndvi_fr_styled_pos <- 
  modis_ndvi_fr_styled  %>% 
  ts_add_positions_to_frames(outline,pcol = "white",psize = 1) %>% 
  ts_add_positions_to_frames(positions = points,
                             position_names = points$Name,
                             tcol = "blue4",
                             t_hjust = -1.5,
                             tsize = 4,
                             psize=3,
                             pcol="blue4",
                             add_text = T,col_by_pos = T)

# rtsVis comes with a variety of plotting functions. We select line2 
# which maps color to position
modis_ndvi_lineframes <-
  modis_ndvi %>%  
  ts_flow_frames(positions = points,
                 position_names = points$Name,
                 pbuffer = 0.1,
                 val_min = -0,
                 val_max = 1,
                 legend_position = "right",
                 plot_function = "line2",
                 aes_by_pos = TRUE,
                 position_legend = TRUE) %>%
  
  # Flow frames too can be adjusted like any ggplot,
  # for instance, to set a specific color scale
  lapply(FUN = function(x){
    x+
      ggtitle("NDVI", "In a 1° radius around the places")+
      ylab("Median NDVI")+
      xlab("Year")+
      theme(aspect.ratio = 0.3)+
      scale_color_brewer("Places",palette = "Set1")
  })
```

Since the positions and the flow frames are thematically connected, it makes sense to combine the two in a single animation.
For this, we again use <a href="https://github.com/16EAGLE/moveVis">`moveVis`</a> functionalities.

Before we do that, we make sure that the colors of the points match those in the graph. Again, existing frames can be easily modified using `lapply` and ggplot2 syntax.

``` r
# apply the same color palette to the spatial frames
modis_ndvi_fr_styled_pos <- modis_ndvi_fr_styled_pos %>% lapply(function(x){
  x+
    scale_color_brewer("Places",palette = "Set1") 
})
  
  

#Join and animate the frames using moveVis functionalities
modis_ndvi_joined <- moveVis::join_frames(
  list(modis_ndvi_fr_styled,
       modis_ndvi_lineframes)
  )

moveVis::animate_frames(modis_ndvi_joined,
                        overwrite = TRUE,
                        out_file = "modis_frames_NDVI.gif",
                        height=525,
                        width=1200,
                        fps = 10,
                        res=75)
  
```
#### Further plot types

`rtsVis`provides a number of preset plot types for visualising multispectral, gradient, or discrete rasters:

* Line plots(`line` and `line2`)

* Violin charts (`vio`)
* Density charts (`dens` and `dens2`)

* Pie charts (`pie`)
* Bar charts (`bar_stack` and `bar_fill`)

They are all used in the same way, by passing the respective argument to `ts_flow_frames`: 

``` r

# Violin frames visualising the distributions of the different bands at different positions
modis_ts_vioframes <-
  modis_ts %>%  
  ts_flow_frames(positions = points[1:3,],
                 position_names = points[1:3,]$Name,
                 pbuffer = 3,
                 legend_position = "right",
                 plot_function = "vio",aes_by_pos = F,
                 position_legend = TRUE,band_colors = c("Red","Purple","Blue","Darkgreen"))


# Density frames visualizing the distributions of the NDVI across positions 
modis_ts_densframes <-
  modis_ndvi %>%  
  ts_flow_frames(positions = points[5:7,],
                 position_names = points[5:7,]$Name,
                 pbuffer = 3,
                 plot_function = "dens2",
                 band_legend_title = "NDVI",
                 band_names = "NDVI",band_colors = c("olivegreen"))

# plenty other types and custom plot functions...

``` 

It is also possible to create custom plot functions and pass them to `ts_flow_frames`, like this:

``` r

custom_plot_function <- function(edf,pl,lp, bl, blt,plt, ps, vs,abp){
    # ... Code for the creation of the ggplot
    # return (plot)
   }
   
modis_ts_vioframes <-
  modis_ts %>%  
  ts_flow_frames(positions = points,
                 position_names = points$Name,
                 pbuffer = 3,
                 plot_function = custom_plot_function)

```



<p align="center"><img src="https://github.com/JohMast/rtsVis_demo/blob/main/Images/modis_frames_NDVI2.gif"></p>

## Demo

For more examples, and a guide on how to create custom plot functions, check out the [Demo](https://github.com/JohMast/rtsVis_demo) repository.

In development, published on CRAN. Last updated: `2021-05-18 17:30:00 CEST`

<p align="center"><img src="https://raw.githubusercontent.com/JohMast/rtsVis_demo/main/Images/NDVI_Northern_Europe_joined_pause.gif"></p>

## Links

To learn more about creating **custom plot types**, access the guide and test material at the associated github [repo](https://github.com/JohMast/rtsVis_demo/).

For creating visualization with movement data, visit [moveVis](http://movevis.org/).

For inspiration, visit the [r-graph-gallery](https://www.r-graph-gallery.com/)!

To learn about data visualization see Claus Wilke's excellent [book](https://clauswilke.com/dataviz/).


NEWS.md
***

## rtsVis 0.0.3
Added additional plot functions to ts_flow_frames.

Improved hillshade support for ts_makeframes.

Various bugfixes.



***



R/data.R
#' Example MODIS time series for southern Slovenia.
#'
#' 105 MODIS MCD43A4 Images covering the time from 2020-02-15 to 2020-05-29 ,
#' downsampled by a factor of 4 and subset to bands 1 to 4
#' to serve as lightweight example time series.
#' Acquired from LP DAAC using the getSpatialData package.
#'
#' @format A list of 105 raster objects
#' @source \url{https://lpdaac.usgs.gov/tools/usgs-earthexplorer/}
"MODIS_SI_ds"



#' Example vector data for Slovenia.
#'
#' Three sets of example vector data for Slovenia. Two sp objects and one matrix object. Chosen to match
#' the "MODIS_SI_ds" raster data.
#' @format A list of three sets of example vector data:
#'  \describe{
#'   \item{points_matrix}{A matrix of coordinates corresponding to the coordinates of two points in Slovenia. Manually created. MODIS sinusoidal projection.}
#'   \item{points}{SpatialPointsDataFrame of four points in Slovenian municipalities. Manually created. MODIS sinusoidal projection.}
#'   \item{polygons}{SpatialPolygonsDataFrame of three polygons corresponding to borders of municipalities. Acquired from GADM. MODIS sinusoidal projection.}
#'   ...
#' }
#' @source \url{https://gadm.org/data.html}
"SI_positions"

R/ts_raster.R
#' Assemble/interpolate a raster time series
#'
#' This function assembles a raster time series by assigning or interpolating input rasters to a target time series.
#'
#' @param r_list a list of raster objects.
#' @param r_times POSIXct, a vector of times corresponding to the elements of \code{r_list}.
#' @param out_times POSIXct, a vector of times for which output rasters will be created.
#' @param fade_raster (Optional) logical. If \code{TRUE} performs a linear interpolation to calculate the values for the output raster. Otherwise uses a nearest temporal neighbor approach. Default is \code{FALSE}. 
#' @param ... additional arguments.
#' @param verbose  (Optional) logical. If \code{TRUE} outputs progress. Default is \code{TRUE}. 
#' @author Jakob Schwalb-Willmann, Johannes Mast
#' @return a list of raster objects.
#' 
#' 
#' @export
#' @examples 
#' #Setup
#' 
#'  library(rtsVis)
#'  # Load example dataset at a greatly increased interval
#' x_list <- MODIS_SI_ds[seq(1,length(MODIS_SI_ds),30)]
#' x_dates <- do.call(c, lapply(MODIS_SI_ds,attr,"time") )[seq(1,length(MODIS_SI_ds),30)]
#' 
#' #Fill NAs
#' x_list_filled <- ts_fill_na(x_list)
#' 
#' #Make a sequence of output dates, double the length of input dates
#' out_dates <-seq.POSIXt(from = x_dates[1],
#'                        to = x_dates[length(x_dates)],length.out = length(x_dates)*2 )
#' 
#' #For each output date, interpolate a raster image from the input files
#' r_list_out <- ts_raster(r_list = x_list_filled,
#'                         r_times = x_dates,
#'                         out_times = out_dates,
#'                         fade_raster = TRUE)
ts_raster <- function(r_list, r_times, out_times = NA, fade_raster = FALSE, ..., verbose = TRUE){
  #2do: take rtype as user input and automatically assign rtype (give warning if auto)
  #2do: take rgb_layers as user input and if more than 3 layers, automatically choose the first 3 as rgb (give warning if auto)
  if(all(!is.list(r_list), inherits(r_list, "Raster"))) r_list <- list(r_list)
  if(length(unique(sapply(r_list, nlayers))) > 1) out("Number of layers per raster object in list 'r' differ.", type = 3)
  if(!inherits(r_times, "POSIXct")) out("Argument 'r_times' must be of type 'POSIXct' if 'r_list' is defined.", type = 3)
  if(!is.logical(fade_raster)) out("Argument 'fade_raster' has to be either TRUE or FALSE.", type = 3)
  if(length(out_times) == 1) out_times <- r_times

  if(identical(r_times, out_times)) return(.ts_set_frametimes(r_list,out_times)) else{
    outr <- .rFrames(r_list, r_times, 
             data.frame(frame = 1:length(out_times), time = sort(out_times)),
             gg.ext = NA, fade_raster, crop_raster = FALSE)
  }
  #add the out dates as attribute "time"
  .ts_set_frametimes(outr,out_times)
  
}

R/ts_fill_na.R
#' Fill NA values in a raster time series
#'
#' @param verbose (Optional) logical. If \code{TRUE} outputs progress. Default is \code{FALSE}. 
#' @param ... additional arguments to be passed on to  \link[raster]{approxNA}. Of particular interest is the \code{rule} argument which defines how first and last cells are dealt with. 
#' @param x_list_fill a list of raster objects.
#' @param maskvalues numeric, a vector of values to be set to NA before the masking.
#'
#' @return A list of rasters with NAs filled.
#' @importFrom raster approxNA
#' @author Johannes Mast
#' @details Loads all layers of a specific bands into a stack and uses \link[raster]{approxNA} to fill the NAs if possible. Note that the procedure requires the entire list of raster layery for each band to be be stacked. It is therefore very memory intensive and likely to fail for very large time series.
#' @export
#' @examples 
#' \donttest{
#' 
#' #Setup
#'  library(rtsVis)
#' x_list <- MODIS_SI_ds[seq(1,length(MODIS_SI_ds),15)]   #A list of raster objects
#' 
#' #Fill NAs
#' x_list_filled <- ts_fill_na(x_list) 
#' }

ts_fill_na <- function(x_list_fill,maskvalues=NULL,verbose=FALSE,...){
  rasternames <- names(x_list_fill)
  n_layers <- nlayers(x_list_fill[[1]])
  for(n_l in 1:n_layers){
    if(verbose){
      print(paste0("Filling Layer:",n_l))
    }
    #make a brick from all the layers
    if(n_layers>1){
      x_lay <- stack( .ts_subset_ts_util(x_list_fill,n_l) )
    }else{
      x_lay <- stack(x_list_fill)
    }
    if(!is.null(maskvalues)){
      print(paste0("Filling Mask values: ", c(maskvalues)))
      x_lay[x_lay %in% maskvalues] <- NA
    }
    
    #fill the nas
    x_lay_filled <- raster::approxNA(x_lay)#,...)
    

    
    #reassign the filled layers to the list elements
    for(n_r in 1:length(x_list_fill)){
      if(n_layers>1){
        x_list_fill[[n_r]][[n_l]] <- x_lay_filled[[n_r]]
      }else{
        x_list_fill <- as.list(x_lay_filled)
      }
    }
  }
  #reassign the original names
  names(x_list_fill) <- rasternames
  return(x_list_fill)
}




R/ts_flow_frames.R
#' Create a series of charts of a raster time series
#'
#' @param r_list list of rasters, as generated by \code{\link{ts_raster}}. 
#' @param positions (Optional) object containing the coordinates. One of \itemize{
##'  \item{A two-column \code{matrix} of coordinates where the first column corresponds to the longitude and the second column corresponds to the latitude.}
##'  \item{A \code{SpatialPolygonsDataFrame} }
##'  \item{A \code{SpatialPointsDataFrame} }
##'  \item{An sf object containing \code{POINTS},\code{POLYGONS} or \code{MULTIPOLYGONS}}
##' }If no positions are specified, one position is assumed to be the entire extent of the rasters.
#' @param position_names (Optional) character, names of the positions to be added in legend or text. By default, will create placeholder names by combining the object type and Id (Example: "Polygon 3")
#' @param val_min (Optional) numeric, minimum value of the y-axis. By default chooses a rounded minimum value of the rasters contained within \code{r_list}.
#' @param val_max (Optional) numeric, maximum value of the y-axis. By default chooses a rounded maximum value of the rasters contained within \code{r_list}.
#' @param val_by (Optional) numeric, interval value of the y-axis. Default is \code{0.1}.
#' @param band_names (Optional) character, names of the bands to be added in legend, if \code{band_legend}. By default, will create placeholder names by band index (Example: "Band 3").
#' @param band_colors (Optional) character, colors which represent the bands in the plot. Must be in matching order to \code{band_names}. If \code{r_list} is discrete, the colors will be mapped to the different levels.  By default will use ggplot colors.
#' @param position_legend (Optional) logical. If \code{TRUE}: Add a legend for the positions. Only recommended if \code{aes_by_pos} is also  \code{TRUE}.
#' @param legend_position  (Optional) character, position of the legend. Use \code{"none"} to disable all legends. Default is \code{"right"}.
#' @param band_legend (Optional) logical. If \code{TRUE}: Add a legend for the bands. Default is \code{TRUE}.
#' @param band_legend_title (Optional) character, title of the band legend. Default is \code{"Bands"}.
#' @param position_legend_title (Optional) character, title of the band legend. Default is \code{"Positions"}.
#' @param pbuffer  (Optional) numeric. The radius of a buffer around each object which will be applied before extraction. By default, no buffer is used.
#' @param plot_function (Optional) character or function, type of the plots to produce. Currently supported are \code{'line'}, \code{'line2'}, \code{'vio'}, \code{'dens'}, \code{'dens2'}, \code{'bar_stack'}, \code{'bar_fill'}, \code{'pie'}.One of \itemize{
##'  \item{\code{'line'} A line chart, suited for comparing trends between bands}
##'  \item{\code{'line2'} A line chart, suited for comparing trends between positions}
##'  \item{\code{'vio'} A density chart, suited for comparing of distributions across bands and positions}
##'  \item{\code{'dens'}  A density chart, suited for comparing of distributions across positions}
##'  \item{\code{'dens2'}  A density chart, suited for comparing of distributions across bands}
##'  \item{\code{'bar_stack'} A horizontal bar chart, suited for visualizing counts and proportions among discrete data.}
##'  \item{\code{'bar_fill'} A horizontal bar chart, suited for visualizing proportions among discrete data.}
##'  \item{\code{'pie'} A pie chart, suited for visualizing rough proportions among discrete data with few categories.}
##' } Alternatively, a custom function with similar structure and arguments can be passed to create other types of plots. Default is \code{"line"}. 
#' @param aes_by_pos (Optional) logical. If \code{TRUE}: vary the linetype aesthetic to be different for each position? If  \code{FALSE}, this also disables the \code{position_legend}, as no notable classes will be plotted. Ignored by some plot types which inherently map position to facets. Default is \code{TRUE}.
#' @param FUN (Optional) function to summarize the values (e.g. mean) during the extraction. See \link[raster]{extract} for more details. Default is \code{"NULL"}. Summarizing in this way is not sensible for many plot types which visualize distribution or count. Note that usually, summarize statistics will be calculated in an appropriate way by the \code{plot_function} rather than during the extraction.
#' @param plot_size (Optional) numeric, size for the ggplot objects. Default is \code{1}.
#' @param return_df (Optional) logical. Return a dataframe with the extracted values instead of a plot? This can be useful for experimenting with plot creation. Default is \code{FALSE}.
#' @param ... (Optional) additional arguments for \code{plot_function}.
#' @details Values are extracted using \link[raster]{extract} and plotted on a \link[ggplot2]{ggplot}.
#'  The type of the ggplot is specified by \code{plot_function}. Currently supported are \code{"line"} and \code{"violin"} as well as custom functions which accept similar inputs.
#'  The function may fail for large polygons and long time series. Be aware that if \code{\link{ts_raster}} is used with \code{fade}, interpolation may be used to generate raster values. 
#' @author Johannes Mast
#' @import sp ggplot2
#' @importFrom raster getValues
#' @importFrom grDevices hcl.colors
#' @importFrom graphics frame
#' @importFrom dplyr left_join group_size group_by
#' @importFrom forcats as_factor
#' @importFrom stats density filter
#' @importFrom magrittr %>%
#' @importFrom tidyr drop_na
#' @return A list of ggplots, one for each element of \code{r_list}.
#' @seealso \code{\link{ts_raster}}
#' @export
#' @examples 
#' #' #Setup
#' library(rtsVis)
#' # Load example dataset at a greatly increased interval
#' x_list <- MODIS_SI_ds[seq(1,length(MODIS_SI_ds),30)]
#' x_dates <- do.call(c, lapply(MODIS_SI_ds,attr,"time") )[seq(1,length(MODIS_SI_ds),30)]
#' 
#' #Fill NAs
#' x_list_filled <- ts_fill_na(x_list)
#' 
#' #Make a sequence of output dates, double the length of input dates
#' out_dates <-seq.POSIXt(from = x_dates[1],
#'                        to = x_dates[length(x_dates)],length.out = length(x_dates)*2 )
#' 
#' #For each output date, interpolate a raster image from the input files
#' r_list_out <- ts_raster(r_list = x_list_filled,
#'                         r_times = x_dates,
#'                         out_times = out_dates,
#'                         fade_raster = TRUE)
#' #Create the frames 
#' # as from the desired layers
#' r_frames <- ts_makeframes(x_list = r_list_out,samplesize = 10,
#'                           l_indices = c(1,4,3))
#' 
#' # Create a line plot from the data extracted over points
#' points <- SI_positions$points #Polygons of Slovenian municipalities covered by the raster
#' flow_frames_point_line <- rtsVis::ts_flow_frames(r_list = r_list_out,
#'  position_names = c("Ljubljana","Ivancna Gorica","Dolenjske Toplice","Loski Potok"),
#'  band_names = c("620 - 670","841 - 876","459 - 479","545 - 565"),
#'  positions = points,
#'  band_colors = c("firebrick3","darkorchid3","dodgerblue3","olivedrab3"),
#'            band_legend_title = "Wavelength [nm]",
#'            position_legend_title = "Obcina",
#'            legend_position = "bottom",
#'            position_legend = FALSE,
#'            band_legend=TRUE,aes_by_pos = TRUE)
#' 
#' #Check one of the frames
#' flow_frames_point_line[[5]]
#' 
#' 
#' # Create a violin plot from the data extracted over polygons
#' # polygons <- SI_positions$polygons
#' #flow_frames_poly_vio <-
#' #rtsVis::ts_flow_frames(r_list = r_list_out,
#' #           position_names = c("Radece","Ljubljana","Kocevje"),
#' #           band_names = c("620 - 670","841 - 876","459 - 479","545 - 565"),
#' #           positions = polygons,
#' #           band_colors = c("firebrick3","darkorchid3","dodgerblue3","olivedrab3"),
#' #           band_legend_title = "Wavelength [nm]",
#' #           position_legend_title = "Obcina",
#' #           position_legend = FALSE,
#' #          legend_position = "left",
#' #           band_legend=TRUE,aes_by_pos = FALSE,
#' #          plot_function = "vio")
#' #Check one of the frames
#' # flow_frames_poly_vio[[5]]
#' 
ts_flow_frames <- function(r_list,positions=NULL,position_names=NULL,band_names=NULL,band_colors=NULL,val_min=NULL,val_max=NULL,val_by=NULL,plot_size=1,position_legend=NULL,legend_position="right",band_legend=NULL,band_legend_title=NULL,position_legend_title=NULL,pbuffer=NULL,plot_function="line",aes_by_pos=TRUE,FUN=NULL,return_df=FALSE,...){
  
  if(class(r_list)!="list"){
    stop(
      "r_list must be a list of raster objects with associated timestamps."
    )
  }
  
  if (is.null(plot_function)) {
    stop(
      "Please provide a plot_function. plot_function must be 'line', 'violin' or an equivalent custom function."
    )
  } else{
    if (class(plot_function) == "character") {
      if(plot_function == "line") {
        print(paste0("Creating: ", length(r_list), " frames of line plots, colored by bands"))
        plot_function <- .ts_gg_line
      } else if(plot_function == "line2") {
        print(paste0("Creating: ", length(r_list), " frames of line plots, colored by position"))
        plot_function <- .ts_gg_line2
      } else if(plot_function == "vio") {
        print(paste0("Creating: ", length(r_list), " frames of violin plots"))
        plot_function <- .ts_gg_vio
      } else if(plot_function == "violin") {
        print(paste0("Creating: ", length(r_list), " frames of violin plots"))
        plot_function <- .ts_gg_vio
      } else if(plot_function == "dens") {
        print(paste0("Creating: ", length(r_list), " frames of density plots, colored by position"))
        plot_function <- .ts_gg_dens
      } else if(plot_function == "dens2"){
        print(paste0("Creating: ", length(r_list), " frames of density plots, colored by band"))
        plot_function <- .ts_gg_dens2
      }else if(plot_function == "bar_stack"){
        print(paste0("Creating: ", length(r_list), " frames of stacked bar plots"))
        plot_function <- .ts_gg_bar_stack
      }else if(plot_function == "bar_fill"){
        print(paste0("Creating: ", length(r_list), " frames of filled bar plots"))
        plot_function <- .ts_gg_bar_fill
      }else if (plot_function == "pie") {
        print(paste0("Creating: ", length(r_list), " frames of pie plots"))
        plot_function <- .ts_gg_pie
      }
    } else if (!class(plot_function) == "function") {
      stop("plot_function must be on of 'line', 'line2', 'vio', 'dens', 'dens2', 'bar_stack', 'bar_fill', 'pie', or an equivalent custom function.")
    }
  }
  if(is.null(plot_function)){
    stop("No suitable plot function found")
  }
  
  #Ensure nice colors
  if(is.null(band_colors)){
    band_colors <-  hcl.colors((nlayers(r_list[[1]])))
  }
  #Ensure nice df names
  if(is.null(band_names)){
    band_names <- paste0("Band",1:(nlayers(r_list[[1]])))  #make artificial bandnames if necessary
  }
  
  
  #Should legends be plotted? By Default no.
  if(is.null(band_legend)){
    if(is.null(band_legend_title)){
      band_legend=FALSE
    }else{
      band_legend=TRUE
    }
  }
  if(is.null(position_legend)){
    if(is.null(position_legend_title)){
      position_legend=FALSE
    }else{
      position_legend=TRUE
    }
  }
  
  
  # Make Default legend titles
  if(is.null(band_legend_title)){
    band_legend_title <- "Band"
  }
  if(is.null(position_legend_title)){
    position_legend_title <- "Positions"
  }
  
  
  
  # extract the values of the raster into a long dataframe
  extract_df <- .ts_extract_from_frames(r_list_extract = r_list,
                                        positions = positions,
                                        position_names = position_names,
                                        band_names = band_names,
                                        pbuffer= pbuffer,
                                        FUN = FUN)
  
  #For most plots we need the data in long format
  if(TRUE){
    extract_df$position_name <- forcats::as_factor(extract_df$position_name)
    extract_df <-  tidyr::pivot_longer(data = extract_df, cols =  band_names,names_to="band") 
  }
  #Make a df to match colors to band names
  color_matching_table <- cbind(band_names,band_colors)
  extract_df <- dplyr::left_join(extract_df,color_matching_table,by = c("band" = "band_names"),copy=TRUE)
  
  ## create value sequence
  if(is.null(val_min)) val_min <- floor_dec(min(sapply(r_list, function(x){min(getValues(x),na.rm = TRUE)})),level = 2)
  if(is.null(val_max)) val_max <- ceiling_dec(max(sapply(r_list, function(x){max(getValues(x),na.rm=TRUE)})),level = 2)
  
  if(is.null(val_by)){
    #If there are no given vals, just make four
    val_seq <- seq(val_min, val_max, length.out = 4)
  }else{
    #Else make the sequence by the given vals
    val_seq <- seq(val_min, val_max, by = val_by)
  }
  
  
  if(return_df){
    return(extract_df)
  }
  
  # flow_frames <- .lapply(1:max(extract_df$frame), function(i){
  #   plot_function(i=i,
  #                 edf = extract_df,
  #                 bl = band_legend,
  #                 pl = position_legend,
  #                 lp = legend_position,
  #                 blt = band_legend_title,
  #                 plt = position_legend_title,
  #                 ps = plot_size,
  #                 vs = val_seq,
  #                 abp = aes_by_pos,...)
  # })
  
  flow_frames <-  plot_function(
    edf = extract_df,
    bl = band_legend,
    pl = position_legend,
    lp = legend_position,
    blt = band_legend_title,
    plt = position_legend_title,
    ps = plot_size,
    vs = val_seq,
    abp = aes_by_pos,
    ...)
  
  
  
  return(flow_frames)
}


